Library/Homebrew/brew.rb

```diff
@@ -166,35 +168,35 @@
       Try to figure out the problem yourself and submit a fix as a pull request.
       We will review it but may or may not accept it.
     EOS
+
   end
 
   exit 1
-rescue RuntimeError, SystemCallError => e
-  raise if e.message.empty?
+rescue Exception => e # rubocop:disable Lint/RescueException
+  runtime_or_system_call_error = e.is_a?(RuntimeError) || e.is_a?(SystemCallError)
+  raise if runtime_or_system_call_error && e.message.empty?
 
   onoe e
-  $stderr.puts Utils::Backtrace.clean(e) if args&.debug? || ARGV.include?("--debug")
 
-  exit 1
-rescue MethodDeprecatedError => e
-  onoe e
-  if e.issues_url
-    $stderr.puts "If reporting this issue please do so at (not Homebrew/brew or Homebrew/homebrew-core):"
-    $stderr.puts "  #{Formatter.url(e.issues_url)}"
+  method_deprecated_error = e.is_a?(MethodDeprecatedError)
+  runtime_or_system_call_or_method_deprecated_error = runtime_or_system_call_error || method_deprecated_error
+  if args&.debug? || ARGV.include?("--debug") || !runtime_or_system_call_or_method_deprecated_error
+    $stderr.puts Utils::Backtrace.clean(e)
   end
-  $stderr.puts Utils::Backtrace.clean(e) if args&.debug? || ARGV.include?("--debug")
-  exit 1
-rescue Exception => e # rubocop:disable Lint/RescueException
-  onoe e
-  if internal_cmd && !OS.unsupported_configuration?
-    if Homebrew::EnvConfig.no_auto_update?
-      $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
-    else
-      $stderr.puts "#{Tty.bold}Please report this issue:#{Tty.reset}"
-      $stderr.puts "  #{Formatter.url(OS::ISSUES_URL)}"
-    end
+
+  if OS.unsupported_configuration?
+    $stderr.puts "#{Tty.bold}Do not report this issue: you are running in an unsupported configuration.#{Tty.reset}"
+  elsif Homebrew::EnvConfig.no_auto_update?
+    $stderr.puts "#{Tty.bold}You have disabled automatic updates.#{Tty.reset}"
+    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
+  elsif (issues_url = (method_deprecated_error && e.issues_url) || Utils::Backtrace.tap_error_url(e))
+    $stderr.puts "If reporting this issue please do so at (not Homebrew/brew or Homebrew/homebrew-core):"
+    $stderr.puts "  #{Formatter.url(issues_url)}"
+  elsif internal_cmd
+    $stderr.puts "#{Tty.bold}Please report this issue:#{Tty.reset}"
+    $stderr.puts "  #{Formatter.url(OS::ISSUES_URL)}"
```

Comment:
It looks like refactor is causing this message to be output more often? Like on a non-existent formula, e.g.
```console
❯ brew uninstall test
Error: No available formula with the name "test". Did you mean testssl?
Please report this issue:
  https://docs.brew.sh/Troubleshooting

❯ git checkout 4.2.16
Note: switching to '4.2.16'.
...

❯ brew uninstall test
Error: No available formula with the name "test". Did you mean testssl?
```

In my opinion, it seems a little odd to tell users to report this.

Reply:
Regression is probably from no longer just exiting on `RuntimeError`. We do classify a bunch of exceptions as `RuntimeError` (in this case, `FormulaUnavailableError < FormulaOrCaskUnavailableError < RuntimeError`).

Reply:
@cho-m Thanks, will take a look.